packages/hydrogen/src/product/useOptimisticVariant.test.tsx

```diff
@@ -1,204 +1,225 @@
-import {expect, test, describe, beforeEach, afterEach, vi} from 'vitest';
+import {expect, it, describe, beforeEach, afterEach, vi} from 'vitest';
+import {act, renderHook} from '@testing-library/react';
+import {useNavigation} from '@remix-run/react';
 import {useOptimisticVariant} from './useOptimisticVariant';
-import {renderHook, waitFor} from '@testing-library/react';
-
-let navigation = {state: 'idle', location: {search: ''}};
 
 describe('useOptimisticVariant', () => {
   beforeEach(() => {
+    vi.stubGlobal('reportError', vi.fn());
     vi.mock('@remix-run/react', async (importOrigninal) => {
       return {
         ...(await importOrigninal<typeof import('@remix-run/react')>()),
-        useNavigation: () => {
-          return navigation;
-        },
+        useNavigation: vi.fn(() => ({state: 'idle', location: {search: ''}})),
       };
     });
-    navigation = {state: 'idle', location: {search: ''}};
-    vi.stubGlobal('reportError', vi.fn());
   });
+
   afterEach(() => {
     vi.clearAllMocks();
   });
-  test('returns the original product if no fetchers are present', () => {
-    const variant = {title: 'Product'};
+
+  it('returns the original product if no fetchers are present', async () => {
+    // Do not create hook params inline inside `renderHook`
+    // to avoid infinite re-renders.
+    const selectedVariant = {title: 'Product'};
+    const variants = {product: {variants: {nodes: []}}};
+
     const {result} = renderHook(() =>
-      useOptimisticVariant(variant, {
-        product: {
-          variants: {nodes: []},
-        },
-      }),
+      useOptimisticVariant(selectedVariant, variants),
```

Comment:
Ah makes sense. I don't think we can `JSON.stringify` the value, because it can also be a promise from remix. But that promise from remix should be stable on re-renders. So I think this is only an issue in the test.